### ISSUE TITLE
[MNG-1525] Allow environment variables to be referenced in pom.xml, settings.xml, etc.

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#3467]
**[Richard Allen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=richard.allen)** opened **[MNG-1525](https://issues.apache.org/jira/browse/MNG-1525?redirect=false)** and commented

Maven should allow environment variables to be referenced; similar to Java System properties. Ant allows this to be done using the following syntax:

    <property environment="env"/>
    ${env.CATALINA_BASE}

where CATALINA_BASE has been defined like:

    export CATALINA_BASE=$HOME/tomcat

In Maven, an example use for signing jars might be:

\<plugin>
\<groupId>org.apache.maven.plugins\</groupId>
\<artifactId>maven-jar-plugin\</artifactId>
\<executions>
\<execution>
\<goals>
\<goal>sign\</goal>
\</goals>
\</execution>
\</executions>
\<configuration>
\<keystore>${env.KEYSTORE_FILE}\</keystore>
\<storepass>${env.KEYSTORE_PASS}\</storepass>
\<alias>${env.KEYSTORE_ALIAS}\</alias>
\<jarPath>${project.build.directory}/${project.build.finalName}.jar\</jarPath>
\</configuration>
\</plugin>

where KEYSTORE_FILE, KEYSTORE_PASS, and KEYSTORE_ALIAS are environment variables.

It would also be very useful to have environment variables available when defining filter properties files. For instance:

1. application.properties
   deploy.dir=${env.CATALINA_BASE}/webapps

Environment variables can be referenced in Java like:

Map<String, String> environmentVariables = java.lang.System.getenv();

---

**Affects:** 2.0

**Attachments:**
- [MNG-1525-maven-project.patch](https://issues.apache.org/jira/secure/attachment/12711604/MNG-1525-maven-project.patch) (_2.15 kB_)

**Issue Links:**
- [MNG-3940](https://issues.apache.org/jira/browse/MNG-3940) Interpolation of environment variables is not case-insensitive on Windows

2 votes, 0 watchers

**Comment by jira-importer** [comment]
**[Edwin Punzalan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elpunzalan)** commented

Can you check if Runtime.exec( "set" ) will print the list of environment variables for your OS ?  Thanks.

**Comment by jira-importer** [comment]
**[Richard Allen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=richard.allen)** commented

No, Runtime.getRuntime().exec("set") does not work on my OS (Red Hat Linux). Java doesn't seem to be able to execute any of the commands that are built into the Bash shell. What would you need to do this for anyway?

**Comment by jira-importer** [comment]
**[Edwin Punzalan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elpunzalan)** commented

You see, Maven's minimum JDK requirement is 1.4.  And getEnv() in 1.4 is already deprecated.

Right now, I'm adding a method to retrieve the environment variables defined in a machine by executing "cmd /C set" for windows, or (since "set" is not working in your OS) "/usr/bin/printenv" for *NIX machines.

We're still open for suggesstions on how to do this, btw. ^_^

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

Edwin, I think it is ok to use getenv() under 1.4. I'm not sure how well it is supported, but we can aim to replace it with the more robust commons-exec later.

**Comment by jira-importer** [comment]
**[Edwin Punzalan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elpunzalan)** commented

Actually, the process I'm looking at is listing all the environment variables, saving it in a Properties object and then using it for interpolation... very much like the way we used it on System.getProperties().  This way, the method would only address machine specific issues which is not really too many since /usr/bin/printenv is common on *NIX.

getEnv() is usable if I am to read pom.xml and extract all ${env*} in it then add it for interpolation which is not really efficient.  Also, not only do I have to read pom.xml but also settings.xml, etc.

**Comment by jira-importer** [comment]
**[Arik Kfir](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arikkfir)** commented

Hi - "System.getenv()" is UNdeprecated under jdk1.5

**Comment by jira-importer** [comment]
**[Richard Allen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=richard.allen)** commented

I've had good success with having Ant recognize environment variables across three different platforms (Linux, Windows, and OS X) with both Java 1.4 and Java 5. Maybe you should look into how Ant reads environment variables. However, since getEnv() isn't deprecated in Java 5, Sun must plan on keeping it around, so I still think that is a viable option. I appreciate your work on this issue.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

commons-exec that I mentioned earlier is derived from the Ant code. When released, we can use that, but getenv might be good enough for now.

**Comment by jira-importer** [comment]
**[skaze](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=skaze)** commented

I guess im going to have to wait for  commons-exec as the only way i know of doing this robustly is via JNI (the codeset used by cmd.exe can be any old crap.)

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

NOT applying this patch. I found a better solution that will factor the interpolation of the POM into a flexible utility in
plexus-utils, and will allow introduction of envar resolution to the POM. It will also make interpolating the settings.xml and profiles.xml using any of a number of expression resolvers (using envar resolution only for now).

BTW, I tried using System.getenv(..) in JDK1.4, and it fails with java.lang.Error and a deprecation message. So, I'm using Edwin's code to extract the envars into a Properties object. We can improve this later.
